File: packages/ui/src/services/crudLog.ts
import { safeInvoke, isTauri, isVSCode } from '@utils/tauri'
import { postVSCodeMessage } from './vscode-bridge'
import type { CrudOperationLog, CrudLogFilter, CrudOperationType } from '@types'

export interface CrudLogStats {
    total: number
    success: number
    failed: number
    inserts: number
    updates: number
    deletes: number
}

export async function addCrudLog(log: CrudOperationLog): Promise<void> {
    if (isVSCode()) return postVSCodeMessage<void>('add_crud_log', { log })
    if (isTauri()) return safeInvoke('add_crud_log', { log }) as Promise<void>
}

export async function queryCrudLogs(
    filter: CrudLogFilter,
    limit: number = 100,
): Promise<CrudOperationLog[]> {
    if (isVSCode()) return postVSCodeMessage<CrudOperationLog[]>('query_crud_logs', { filter, limit })
    if (isTauri()) return safeInvoke('query_crud_logs', { filter, limit }) as Promise<CrudOperationLog[]>
    return []
}

export async function countCrudLogsByTab(tabId: string): Promise<number> {
    if (isVSCode()) return postVSCodeMessage<number>('count_crud_logs_by_tab', { tabId })
    if (isTauri()) return safeInvoke('count_crud_logs_by_tab', { tabId }) as Promise<number>
    return 0
}

export async function deleteCrudLogsByTab(tabId: string): Promise<number> {
    if (isVSCode()) return postVSCodeMessage<number>('delete_crud_logs_by_tab', { tabId })
    if (isTauri()) return safeInvoke('delete_crud_logs_by_tab', { tabId }) as Promise<number>
    return 0
}

export async function getCrudLogTableNames(): Promise<string[]> {
    if (isVSCode()) return postVSCodeMessage<string[]>('get_crud_log_table_names')
    if (isTauri()) return safeInvoke('get_crud_log_table_names') as Promise<string[]>
    return []
}

export async function getCrudLogStats(connectionId?: string): Promise<CrudLogStats> {
    if (isVSCode()) return postVSCodeMessage<CrudLogStats>('get_crud_log_stats', { connectionId })
    if (isTauri()) return safeInvoke('get_crud_log_stats', { connectionId }) as Promise<CrudLogStats>
    return { total: 0, success: 0, failed: 0, inserts: 0, updates: 0, deletes: 0 }
}

export function createCrudLog(
    connectionId: string,
    tabId: string,
    tableName: string,
    operationType: CrudOperationType,
    sql: string,
): CrudOperationLog {
    return {
        id: crypto.randomUUID(),
        connection_id: connectionId,
        tab_id: tabId,
        table_name: tableName,
        operation_type: operationType,
        sql,
        row_data: undefined,
        old_data: undefined,
        rows_affected: 0,
        executed_at: new Date().toISOString(),
        duration_ms: 0,
        is_success: true,
        error_message: undefined,
    }
}

File: apps/desktop/src-tauri/src/commands/crud_log.rs
use std::sync::Arc;

use tauri::State;

use crate::core::crud_log_store::CrudLogStore;
use crate::models::crud_log::{CrudLogFilter, CrudOperationLog, CrudOperationType};
use crate::utils::error::AppResult;

/// 添加 CRUD 操作日志
#[tauri::command]
pub async fn add_crud_log(
    log_store: State<'_, Arc<CrudLogStore>>,
    log: CrudOperationLog,
) -> AppResult<()> {
    log_store.add_log(&log)
}

/// 查询 CRUD 操作日志
#[tauri::command]
pub async fn query_crud_logs(
    log_store: State<'_, Arc<CrudLogStore>>,
    filter: CrudLogFilter,
    limit: usize,
) -> AppResult<Vec<CrudOperationLog>> {
    log_store.query_logs(&filter, limit)
}

/// 获取指定 Tab 的日志数量
#[tauri::command]
pub async fn count_crud_logs_by_tab(
    log_store: State<'_, Arc<CrudLogStore>>,
    tab_id: String,
) -> AppResult<usize> {
    log_store.count_logs_by_tab(&tab_id)
}

/// 删除指定 Tab 的日志
#[tauri::command]
pub async fn delete_crud_logs_by_tab(
    log_store: State<'_, Arc<CrudLogStore>>,
    tab_id: String,
) -> AppResult<usize> {
    log_store.delete_logs_by_tab(&tab_id)
}

/// 获取所有表名（用于筛选）
#[tauri::command]
pub async fn get_crud_log_table_names(
    log_store: State<'_, Arc<CrudLogStore>>,
) -> AppResult<Vec<String>> {
    log_store.get_table_names()
}

/// 获取 CRUD 日志统计
#[tauri::command]
pub async fn get_crud_log_stats(
    log_store: State<'_, Arc<CrudLogStore>>,
    connection_id: Option<String>,
) -> AppResult<CrudLogStatsResponse> {
    let stats = log_store.get_stats(connection_id.as_deref())?;
    Ok(CrudLogStatsResponse {
        total: stats.total,
        success: stats.success,
        failed: stats.total - stats.success,
        inserts: stats.inserts,
        updates: stats.updates,
        deletes: stats.deletes,
    })
}

/// 统计信息响应
#[derive(serde::Serialize)]
pub struct CrudLogStatsResponse {
    pub total: usize,
    pub success: usize,
    pub failed: usize,
    pub inserts: usize,
    pub updates: usize,
    pub deletes: usize,
}
